src/devicecontroller/DefaultDeviceController.ts

```diff
@@ -293,131 +203,71 @@ export default class DefaultDeviceController
-     *
-     * If you are confident that your application does not use AudioWorklet, does not run in
-     * an un-fixed Chromium version, or will never be used with sample-rate-switching Bluetooth
-     * devices, you can disable this workaround by suppling a custom {@link ExtendedBrowserBehavior}
-     * when you create your device controller.
-     *
-     * We can't tell in advance whether we need to give the device a different audio context,
-     * because checking whether the resulting node is an AudioWorkletNode needs it to have been
-     * created first.
-     */
-
-    // Ideally we would only do this work if we knew the device was going to change.
-
-    // By definition, this is only needed for Web Audio.
-    let recreateAudioContext = this.useWebAudio;
-
-    if (!this.useWebAudio) {
-      this.logger.debug('Not using Web Audio. No need to recreate audio context.');
-    }
-
-    // We have a suspended audio context. There's nothing we can do, and there's
-    // certainly no point in recreating it. Choosing the transform device will try to resume.
-    if (DefaultDeviceController.audioContext?.state === 'suspended') {
-      recreateAudioContext = false;
-    }
-
-    // Only Chrome needs this fix.
-    if (recreateAudioContext && !this.browserBehavior.requiresContextRecreationForAudioWorklet()) {
-      this.logger.debug('Browser does not require audio context recreation hack.');
-      recreateAudioContext = false;
+      return undefined;
     }
-
-    // Only need to do this if either device has an audio worklet.
-    if (recreateAudioContext && !this.transform && isAudioTransformDevice(device)) {
-      this.logger.debug('Neither device is a transform. No need to recreate audio context.');
-      recreateAudioContext = false;
+    if (this.getAudioInputInProgress) {
+      throw new Error('Another start audio input is in progress');
     }
+    this.getAudioInputInProgress = true;
 
-    if (recreateAudioContext) {
-      this.logger.info('Recreating audio context when selecting new device.');
-
-      /* istanbul ignore else */
-      if (this.transform) {
-        /* istanbul ignore else */
-        if (this.transform.nodes) {
-          this.transform.nodes.end.disconnect();
-          this.transform.nodes = undefined;
-        }
-        this.transform = undefined;
-      }
+    try {
+      if (isAudioTransformDevice(device)) {
+        // N.B., do not JSON.stringify here — for some kinds of devices this
+        // will cause a cyclic object reference error.
+        this.logger.info(`Choosing transform input device ${device}`);
 
-      /* istanbul ignore else */
-      if (this.audioInputSourceNode) {
-        this.audioInputSourceNode.disconnect();
-        this.audioInputSourceNode = undefined;
+        await this.chooseAudioTransformInputDevice(device);
+      } else {
+        this.logger.info(`Choosing intrinsic audio input device ${device}`);
+        this.removeTransform();
```

Comment:
Should we trigger this only on whether the existing selected device is a transform one?